phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java

```diff
@@ -0,0 +1,422 @@
+            int columnIndex = 0;
+            while (!columns.isEmpty() && columnIndex < columns.size()) {
+                if (columns.get(columnIndex).isEmpty()) {
+                    columns.remove(columnIndex);
+                } else {
+                    columnIndex++;
+                }
+            }
+        }
+        return true;
+    }
+
+    private void addCells(CompactionRowVersion rowVersion, List<Cell> result) {
+        if (rowVersion.columnDeleteMarkers == null) {
+            result.addAll(rowVersion.cells);
+            return;
+        }
+        for (Cell cell : rowVersion.cells) {
+            if (rowVersion.columnDeleteMarkers.isEmpty()) {
+                result.add(cell);
+            }
+            else {
+                Cell dm = rowVersion.columnDeleteMarkers.get(0);
+                if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(),
+                        cell.getQualifierLength(), dm.getQualifierArray(), dm.getQualifierOffset(),
+                        dm.getQualifierLength()) == 0) {
+                    rowVersion.columnDeleteMarkers.remove(0);
+                    continue;
+                }
+                else {
+                    result.add(cell);
+                }
+            }
+        }
+    }
+    private void updateResultsWithRowOutsideMaxLookbackButInsideTTLWindow(List<Cell> result,
+            CompactionRowVersion rowVersion) {
+        // The rows that are outside the max lookback window but inside the TTL window
+        if (rowVersion.deleteFamilyMarker == null) {
+            if (rowVersion.version < maxVersion) {
+                // Live rows whose version is less than max version are retained inside
+                // the TTL window
+                addCells(rowVersion, result);
+            }
+        }
+        else if (keepDeletedCells == KeepDeletedCells.TTL) {
+            // Deleted rows inside the TTL window with KeepDeletedCells.TTL are retained
+            addCells(rowVersion, result);
+        }
+        else if (keepDeletedCells == KeepDeletedCells.TRUE) {
+            if (rowVersion.version < maxVersion) {
+                // Deleted rows inside the TTL window with KeepDeletedCells.TRUE are
+                // retained if their version is less than max version
+                addCells(rowVersion, result);
+            }
+        }
+    }
+
+    private void updateResultsWithRowOutsideTTLWindow(List<Cell> result,
+            CompactionRowVersion rowVersion) {
+        if (rowVersion.deleteFamilyMarker == null) {
+            if (firstStore) {
+                if (rowVersion.version < minVersion) {
+                    // Live rows whose version is less than min version are retained outside
+                    // the TTL window
+                    addCells(rowVersion, result);
+                }
+            } else if (rowVersion.version < maxVersion) {
+                // check if this row needs to be retained
```

Comment:
no logic here -- the check is elsewhere? If so please mention

Reply:
I moved the logic for detecting the need for a region level scan to formRowVersions. I will update the code.